test/Language/Haskell/ExpressionSpec.hs

```diff
@@ -0,0 +1,85 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Language.Haskell.ExpressionSpec (spec) where
+
+import           Data.Char
+import           Test.Hspec
+import           Test.QuickCheck
+import qualified Language.Haskell.Meta.Parse as M
+import           Language.Haskell.TH.Syntax
+import           Language.Haskell.TH.Ppr
+
+import           Language.Haskell.Expression
+
+
+spec :: Spec
+spec = do
+
+  describe "parseExpression" $ do
+
+    context "variables" $ do
+      it "parses variables" $ testEqExp "foo"
+
+    context "literal" $ do
+      it "parses Char literals" $ testEqExp "'c'"
+      it "parses String literals" $ testEqExp "\"str\""
+      it "parses Integer literals" $ testEqExp "5"
+
+    context "application" $ do
+      it "parses function application" $ testEqExp "f x"
+
+    context "type signatures" $ do
+      it "accepts variables with type signatures" $ testEqExp "foo :: Int"
+
+    it "parses arbitrary expressions" $ do
+      property $ \x -> testEqExpR . show $ ppr (x :: Exp)
+
+
+testEqExp :: String -> Expectation
+testEqExp e = parseExp e `shouldBe` M.parseExp e
+
+-- | Doesn't check the Left case for equality
+testEqExpR ::String -> Expectation
+testEqExpR e = toMaybe (parseExp e) `shouldBe` toMaybe (M.parseExp e)
+  where toMaybe = either (const Nothing) Just
+
+-- * QuickCheck
+
+-- Type for variable names
+newtype LowerName = LowerName { unLowerName :: Name }
+  deriving (Eq, Show)
+
+-- Type for constructor names
+newtype UpperName = UpperName { unUpperName :: Name }
+  deriving (Eq, Show)
+
+
+validIdentLetter :: Char -> Bool
+validIdentLetter x = x == '_' || x == '\'' || isLetter x
+
+instance Arbitrary LowerName where
+  arbitrary = LowerName . mkName <$> arbitrary `suchThat` go
+    where go []     = False
+          go (x:xs) = isAsciiLower x && isLetter x && all validIdentLetter xs
+
+instance Arbitrary UpperName where
+  arbitrary = UpperName . mkName <$> arbitrary `suchThat` go
+    where go []     = False
+          go (x:xs) = isAsciiUpper x && isLetter x && all validIdentLetter xs
+
+
+instance Arbitrary Lit where
+  arbitrary = oneof [ CharL <$> arbitrary
+                    , StringL <$> arbitrary
+                    , IntegerL <$> arbitrary `suchThat` (>= 0)
+                    ]
+
+instance Arbitrary Type where
+  arbitrary = oneof [ ConT . unUpperName <$> arbitrary
+                    , VarT . unLowerName <$> arbitrary
+                    ]
+
+instance Arbitrary Exp where
+  arbitrary = oneof [ VarE . unLowerName <$> arbitrary
+                    , LitE <$> arbitrary
+                    , SigE <$> arbitrary <*> arbitrary
+                    ]
```

Comment:
Wouldn't it be much simpler to have a `genLowerName :: Gen Name` instead of the newtype wrapper and the `Arbitrary` instance? (Same for `genUpperName`, of course.)